File: archive/j/java/LongestCommonSubsequence.java
import java.util.*;

public class LongestCommonSubsequence {

    static ArrayList<String> split_strings(String str) {
        String word = "";
        char dl = ',';

        str = str + dl;

        ArrayList<String> substr_list = new ArrayList<>();
        for (int i = 0; i < str.length(); i++) {

            if (str.charAt(i) != dl)
                word = word + str.charAt(i);
            else {
                if ((int) word.length() != 0)
                    substr_list.add(word);
                word = "";
            }
        }

        return substr_list;
    }

    static void longest_common_subsequence(ArrayList<String> arr1, ArrayList<String> arr2) {

        int[][] array = new int[arr1.size() + 1][arr2.size() + 1];

        for (int i = 0; i < array.length; i++) {
            for (int j = 0; j < array[0].length; j++) {

                if (i == 0 || j == 0) {
                    array[i][j] = 0;
                } else if (arr1.get(i - 1).equals(arr2.get(j - 1))) {
                    array[i][j] = 1 + array[i - 1][j - 1];
                } else {
                    array[i][j] = Math.max(array[i - 1][j], array[i][j - 1]);
                }
            }
        }

        ArrayList<String> lcs = new ArrayList<>();
        int x = array.length - 1;
        int y = array[0].length - 1;

        while (x > 0 && y > 0) {

            if (arr1.get(x - 1).equals(arr2.get(y - 1))) {
                lcs.add(0, arr1.get(x - 1));
                x--;
                y--;
            } else {
                if (array[x - 1][y] > array[x][y - 1])
                    x--;
                else
                    y--;
            }
        }

        for (int element = 0; element < lcs.size() - 1; element++) {
            System.out.print(lcs.get(element) + ",");
        }
        System.out.println(lcs.get(lcs.size() - 1));
    }

    public static void main(String args[]) {

        if (args.length < 2 || args[0] == "" || args[1] == "") {
            System.out.println("Usage: please provide two lists in the format \"1, 2, 3, 4, 5\"");
            return;
        }

        ArrayList<String> arr1 = split_strings(args[0]);
        ArrayList<String> arr2 = split_strings(args[1]);

        longest_common_subsequence(arr1, arr2);
    }
}

File: archive/t/typescript/longest-common-subsequence.ts
"use strict";

function printUsage(): void {
  console.error(
    'Usage: please provide two lists in the format "1, 2, 3, 4, 5"',
  );
  process.exit(1);
}

function parseInput(input: string | undefined): number[] | null {
  if (!input || input.trim() === "") return null;

  const parts = input.split(",").map((s) => s.trim());
  const arr: number[] = [];

  for (const p of parts) {
    if (p === "") return null;

    const val = Number.parseInt(p, 10);
    if (Number.isNaN(val)) return null;

    arr.push(val);
  }

  return arr;
}

function lcs(arr1: number[], arr2: number[]): void {
  const n = arr1.length;
  const m = arr2.length;

  const dp: number[][] = Array.from({ length: n + 1 }, () =>
    Array(m + 1).fill(0),
  );

  for (let i = 1; i <= n; i++) {
    for (let j = 1; j <= m; j++) {
      if (arr1[i - 1] === arr2[j - 1]) {
        dp[i][j] = 1 + dp[i - 1][j - 1];
      } else {
        dp[i][j] = Math.max(dp[i - 1][j], dp[i][j - 1]);
      }
    }
  }

  const result: number[] = [];
  let i = n;
  let j = m;

  while (i > 0 && j > 0) {
    if (arr1[i - 1] === arr2[j - 1]) {
      result.push(arr1[i - 1]);
      i--;
      j--;
    } else if (dp[i - 1][j] >= dp[i][j - 1]) {
      i--;
    } else {
      j--;
    }
  }

  result.reverse();
  console.log(result.join(", "));
}

function main(): void {
  const input1 = process.argv[2];
  const input2 = process.argv[3];

  const arr1 = parseInput(input1);
  const arr2 = parseInput(input2);

  if (!arr1 || !arr2) {
    printUsage();
  }

  lcs(arr1!, arr2!);
}

main();
